flink-table/flink-sql-parser-hive/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -218,4 +218,549 @@ SqlNode TableOption() :
+    {
+        return new SqlDataTypeSpec(typeName, s.end(this)).withNullable(nullable);
+    }
+}
+
+HiveTableStoredAs TableStoredAs(SqlParserPos pos) :
+{
+  SqlIdentifier fileFormat = null;
+  SqlCharStringLiteral inputFormat = null;
+  SqlCharStringLiteral outputFormat = null;
+}
+{
+  (
+    LOOKAHEAD(2)
+    <INPUTFORMAT> <QUOTED_STRING> { inputFormat = createStringLiteral(token.image, getPos()); }
+    <OUTPUTFORMAT> <QUOTED_STRING> { outputFormat = createStringLiteral(token.image, getPos()); }
+    { return HiveTableStoredAs.ofInputOutputFormat(pos, inputFormat, outputFormat); }
+    |
+    fileFormat = SimpleIdentifier()
+    { return HiveTableStoredAs.ofFileFormat(pos, fileFormat); }
+  )
+}
+
+HiveTableRowFormat TableRowFormat(SqlParserPos pos) :
+{
+  SqlCharStringLiteral fieldsTerminator = null;
+  SqlCharStringLiteral escape = null;
+  SqlCharStringLiteral collectionTerminator = null;
+  SqlCharStringLiteral mapKeyTerminator = null;
+  SqlCharStringLiteral linesTerminator = null;
+  SqlCharStringLiteral nullAs = null;
+  SqlCharStringLiteral serdeClass = null;
+  SqlNodeList serdeProps = null;
+}
+{
+  (
+    <DELIMITED>
+      [ <FIELDS> <TERMINATED> <BY> <QUOTED_STRING>
+        { fieldsTerminator = createStringLiteral(token.image, getPos()); }
+        [ <ESCAPED> <BY> <QUOTED_STRING> { escape = createStringLiteral(token.image, getPos()); } ]
+      ]
+      [ <COLLECTION> <ITEMS> <TERMINATED> <BY> <QUOTED_STRING> { collectionTerminator = createStringLiteral(token.image, getPos()); } ]
+      [ <MAP> <KEYS> <TERMINATED> <BY> <QUOTED_STRING> { mapKeyTerminator = createStringLiteral(token.image, getPos()); } ]
+      [ <LINES> <TERMINATED> <BY> <QUOTED_STRING> { linesTerminator = createStringLiteral(token.image, getPos()); } ]
+      [ <NULL> <DEFINED> <AS> <QUOTED_STRING> { nullAs = createStringLiteral(token.image, getPos()); } ]
+      { return HiveTableRowFormat.withDelimited(pos, fieldsTerminator, escape, collectionTerminator, mapKeyTerminator, linesTerminator, nullAs); }
+    |
+    <SERDE> <QUOTED_STRING>
+    {
+      serdeClass = createStringLiteral(token.image, getPos());
+    }
+    [ <WITH> <SERDEPROPERTIES> serdeProps = TableProperties() ]
+    { return HiveTableRowFormat.withSerDe(pos, serdeClass, serdeProps); }
+  )
+}
+
+/**
+* A sql type name extended basic data type, it has a counterpart basic
+* sql type name but always represents as a special alias compared with the standard name.
+*
+* <p>For example, STRING is synonym of VARCHAR(INT_MAX)
+* and BYTES is synonym of VARBINARY(INT_MAX).
+*/
+SqlTypeNameSpec ExtendedSqlBasicTypeName() :
+{
+    final SqlTypeName typeName;
+    final String typeAlias;
+    int precision = -1;
+}
+{
+    (
+        <STRING> {
+            typeName = SqlTypeName.VARCHAR;
+            typeAlias = token.image;
+            precision = Integer.MAX_VALUE;
+        }
+    |
+        <BYTES> {
+            typeName = SqlTypeName.VARBINARY;
+            typeAlias = token.image;
```

Comment:
It seems that Hive does not support `BYTES` type [1] and its varchar type max precision is 65535.
[1] https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-CreateTableCreate/Drop/TruncateTable
[2] https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types#LanguageManualTypes-ColumnTypes

Reply:
Nice catch about the `BYTES` type, will remove it.
We don't have to worry about varchar in the parser. HiveCatalog will error out if the precision exceeds 65535.

Reply:
I mean for default precision `VARCHAR` in Hive, the precision should be 655235 instead of `Integer.MAX_INT`.

Reply:
Hive doesn't have a default precision for `VARCHAR`. If users don't specify a precision in Hive, they'll hit an error. So I guess it's OK to hit an error in Flink too.

Reply:
I have updated to disallow specifying varchar w/o a precision. It doesn't check whether precision is less than 65535, because we need `VARCHAR(Integer.MAX_INT)` to represent `STRING`

Reply:
What is the `STRING` type precision in Hive ?

Reply:
It's Integer.MAX_INT